### PR TITLE
UCP/CORE: Fixed wiface activation counter.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Added usage counter to ep configuration structure to track the number of endpoints using the configuration.
Reverted https://github.com/openucx/ucx/pull/9104.
Activated UCP worker interfaces when setting endpoint configuration to an endpoint.

## Why ?
The changes are required to properly deactivate worker interface when it is unused.
Currently we may deactivate worker interface even if there are still active endpoints.

### CODE REVIEW

File: src/ucp/core/ucp_worker.h [hunk 1]
```diff
@@ -244,6 +244,7 @@ struct ucp_worker_iface {
     int                           event_fd;      /* Event FD, or -1 if undefined */
     unsigned                      activate_count;/* How many times this iface has
```
Comment (rakhmets) [inline src/ucp/core/ucp_worker.h:245]: Rename to `use_count`.

File: src/ucp/core/ucp_ep.h [hunk 2]
```diff
@@ -885,4 +891,21 @@ ucs_status_t ucp_ep_query_sockaddr(ucp_ep_h ucp_ep, ucp_ep_attr_t *attr);
  */
 ucs_status_t ucp_ep_realloc_lanes(ucp_ep_h ep, unsigned new_num_lanes);
 
+
+/**
+ * Activate UCP worker interfaces corresponding to the lines selected by the
```
Comment (brminich) [inline src/ucp/core/ucp_ep.h:896]:
```suggestion
 * Activate UCP worker interfaces corresponding to the lanes selected by the
```

File: src/ucp/core/ucp_ep.h [hunk 3]
```diff
@@ -885,4 +891,21 @@ ucs_status_t ucp_ep_query_sockaddr(ucp_ep_h ucp_ep, ucp_ep_attr_t *attr);
...
+ * protocols.
+ *
+ * @param [in]  ep  Endpoint object.
+ */
+void ucp_ep_activate_worker_ifaces(ucp_ep_h ep);
+
+/**
+ * Deactivate UCP worker interfaces corresponding to the lines selected by the
```
Comment (brminich) [inline src/ucp/core/ucp_ep.h:904]:
```suggestion
 * Deactivate UCP worker interfaces corresponding to the lanes selected by the
```

File: src/ucp/wireup/wireup_ep.c [hunk 4]
```diff
@@ -271,16 +271,23 @@ void ucp_wireup_ep_discard_aux_ep(ucp_wireup_ep_t *wireup_ep,
...
 
     ucp_wireup_ep_disown(&wireup_ep->super.super, aux_ep);
-    ucp_worker_discard_uct_ep(ucp_ep, aux_ep, wireup_ep->aux_rsc_index,
-                              ep_flush_flags, purge_cb, purge_arg,
+    rsc_index = wireup_ep->aux_rsc_index;
+    ucp_worker_discard_uct_ep(ucp_ep, aux_ep, rsc_index, ep_flush_flags,
+                              purge_cb, purge_arg,
                               (ucp_send_nbx_callback_t)ucs_empty_function,
                               NULL);
+    worker = ucp_ep->worker;
```
Comment (brminich) [inline src/ucp/wireup/wireup_ep.c:287]: minor: can init in place

File: src/ucp/core/ucp_ep.c [hunk 5]
```diff
@@ -796,6 +796,7 @@ ucs_status_t ucp_ep_init_create_wireup(ucp_ep_h ep, unsigned ep_init_flags,
         return status;
     }
 
+    ucp_ep_activate_worker_ifaces(ep);
```
Comment (brminich) [inline src/ucp/core/ucp_ep.c:799]: can be it done in `ucp_worker_get_ep_config` somehow?

File: src/ucp/wireup/wireup_cm.c [hunk 6]
```diff
@@ -510,6 +510,7 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
             goto err;
         }
 
+        ucp_ep_activate_worker_ifaces(ep);
```
Comment (brminich) [inline src/ucp/wireup/wireup_cm.c:513]: can it be done in `ucp_worker_get_ep_config` instead?

File: src/ucp/core/ucp_ep.c (hunk 5 above)
Comment (rakhmets) [reply src/ucp/core/ucp_ep.c:799]: `ucp_worker_get_ep_config` gets the configuration. However, it doesn't set the configuration to the endpoint. `ucp_ep_activate_worker_ifaces` increases the number of endpoints using the endpoint configuration. Thus it should be only called when the configuration is set to the endpoint.

File: src/ucp/wireup/wireup_ep.c (hunk 4 above)
Comment (rakhmets) [reply src/ucp/wireup/wireup_ep.c]: Fixed.

File: src/ucp/core/ucp_ep.c (hunk 5 above)
Comment (brminich) [reply src/ucp/core/ucp_ep.c:799]: imo increasing `ep_count` of `ep_config` in `ucp_ep_activate_worker_ifaces` is confusing. Maybe it is worth to split `ucp_ep_activate_worker_ifaces` and increase `ep_count` in `ucp_worker_get_ep_config`?

Comment (rakhmets) [reply src/ucp/core/ucp_ep.c:799]: But calling `ucp_worker_get_ep_config` doesn't mean that endpoint will use the configuration. It is not correct to increase the counter inside the `ucp_worker_get_ep_config`.

Comment (rakhmets) [reply src/ucp/core/ucp_ep.c:799]: We can do the following instead:
```c
...
    get_ep_config(&cfg_idx);
    set_ep_config(ep, cfg_idx);
...

set_ep_config(ep, cfg_idx)
{
    ep->cfg_idx = cfg_idx;
    increase_cnt(cfg_idx);
    activate_ifaces()
}
```

File: src/ucp/core/ucp_ep.c [hunk 7]
```diff
@@ -3771,3 +3775,44 @@ ucs_status_t ucp_ep_realloc_lanes(ucp_ep_h ep, unsigned new_num_lanes)
 
     return UCS_OK;
 }
+
+static void
+ucp_ep_for_each_proto_lane(ucp_ep_h ep,
+                           void wiface_process(ucp_worker_iface_t *))
```
Comment (iyastreb) [inline src/ucp/core/ucp_ep.c:3781]: Minor: for function pointer I used to see `*`, strange that it works without:
```
void (*wiface_process)(ucp_worker_iface_t *)
```

File: src/ucp/core/ucp_ep.c [hunk 8]
```diff
@@ -3771,3 +3775,44 @@ ucs_status_t ucp_ep_realloc_lanes(ucp_ep_h ep, unsigned new_num_lanes)
...
+    }
+}
+
+void ucp_ep_activate_worker_ifaces(ucp_ep_h ep)
+{
+    ucp_ep_config_t *ep_config = ucp_ep_config(ep);
+
+    ucs_trace("activate wifaces ep %p config index %u ep count %u", ep,
+              ep->cfg_index, ep_config->ep_count);
+    if (ep_config->ep_count++ > 0) {
```
Comment (iyastreb) [inline src/ucp/core/ucp_ep.c:3798]: Minor: maybe shorter version is more readable:
```
if (0 == ep_config->ep_count++) {
    ucp_ep_for_each_proto_lane(ep, ucp_worker_iface_progress_ep);
}

File: src/ucp/core/ucp_ep.c [hunk 9]
```diff
@@ -3771,3 +3775,44 @@ ucs_status_t ucp_ep_realloc_lanes(ucp_ep_h ep, unsigned new_num_lanes)
...
+
+void ucp_ep_deactivate_worker_ifaces(ucp_ep_h ep)
+{
+    ucp_ep_config_t *ep_config = ucp_ep_config(ep);
+
+    ucs_trace("deactivate wifaces ep %p config index %u ep count %u", ep,
+              ep->cfg_index, ep_config->ep_count);
+    ucs_assertv(ep_config->ep_count > 0, "ep %p config index %u", ep,
+                ep->cfg_index);
+    if (--ep_config->ep_count > 0) {
```
Comment (iyastreb) [inline src/ucp/core/ucp_ep.c:3813]: minor: same here?

File: src/ucp/proto/proto_select.c [hunk 10]
```diff
@@ -551,20 +551,22 @@ ucp_proto_select_wiface_activate(ucp_worker_h worker,
...
+    ucp_ep_config_t *ep_config;
     ucp_lane_index_t lane;
     ucp_rsc_index_t rsc_index;
     ucp_worker_iface_t *wiface;
 
-    lane_map      = ucp_proto_select_get_lane_map(worker, select_elem);
-    ep_config_key = &ucs_array_elem(&worker->ep_config, ep_cfg_index).key;
+    lane_map  = ucp_proto_select_get_lane_map(worker, select_elem);
+    ep_config = &ucs_array_elem(&worker->ep_config, ep_cfg_index);
+    lane_map &= ~ep_config->proto_lane_map;
```
Comment (yosefe) [inline src/ucp/proto/proto_select.c:561]:
```c
ep_config = &ucs_array_elem(&worker->ep_config, ep_cfg_index);
lane_map  = ucp_proto_select_get_lane_map(worker, select_elem) & 
            ~ep_config->proto_lane_map;
```

File: src/ucp/proto/proto_select.c [hunk 11]
```diff
@@ -551,20 +551,22 @@ ucp_proto_select_wiface_activate(ucp_worker_h worker,
...
     ucs_for_each_bit(lane, lane_map) {
         ucs_assertv(lane < UCP_MAX_LANES,
                     "lane=%" PRIu8 ", lane_map=0x%" PRIx16, lane, lane_map);
-        rsc_index = ep_config_key->lanes[lane].rsc_index;
+        rsc_index = ep_config->key.lanes[lane].rsc_index;
         wiface    = ucp_worker_iface(worker, rsc_index);
         ucp_worker_iface_progress_ep(wiface);
-        wiface->flags |= UCP_WORKER_IFACE_FLAG_KEEP_ACTIVE;
+
+        ep_config->proto_lane_map |= UCS_BIT(lane);
```
Comment (yosefe) [inline src/ucp/proto/proto_select.c:569]: we can do after the loop: `ep_config->proto_lane_map |= lane_map`

File: src/ucp/core/ucp_ep.c (hunk 5 above)
Comment (yosefe) [reply src/ucp/core/ucp_ep.c:799]: yeah, we can add new function ucp_ep_set_cfg_index(ep, new_cfg_index) - that will deactivate according to the exisitng ep->cfg_index (if it is not NULL) and activate according to the new cfg_index

File: src/ucp/core/ucp_ep.c [hunk 12]
```diff
@@ -1558,6 +1559,48 @@ void ucp_ep_set_failed_schedule(ucp_ep_h ucp_ep, ucp_lane_index_t lane,
...
+
+static void
+ucp_ep_config_deactivate_worker_ifaces(ucp_worker_h worker,
+                                       ucp_worker_cfg_index_t cfg_index)
+{
+    ucp_ep_config_t *ep_config = ucp_worker_ep_config(worker, cfg_index);
+
+    ucs_trace("deactivate wifaces worker %p ep config %u ep count %u", worker,
+              cfg_index, ep_config->ep_count);
+    if (--ep_config->ep_count == 0) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:1598]: maybe assert ep_config->ep_count > 0 before this?

File: src/ucp/core/ucp_worker.inl [hunk 13]
```diff
@@ -118,6 +118,20 @@ ucp_worker_iface(ucp_worker_h worker, ucp_rsc_index_t rsc_index)
                                                                 rsc_index)];
 }
 
+/**
+ * @return worker-iface struct by ep configuration and lane index
+ */
+static UCS_F_ALWAYS_INLINE ucp_worker_iface_t*
+ucp_worker_iface_by_lane(ucp_worker_h worker, ucp_ep_config_t *ep_config,
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.inl:125]: ep_config can be const

**Review by yosefe — APPROVED** [review]

File: src/ucp/proto/proto_select.c [hunk 14]
```diff
@@ -550,22 +550,22 @@ ucp_proto_select_wiface_activate(ucp_worker_h worker,
...
-    ep_config_key = &ucs_array_elem(&worker->ep_config, ep_cfg_index).key;
+    ep_config = ucp_worker_ep_config(worker, ep_cfg_index);
+    lane_map  = ucp_proto_select_get_lane_map(worker, select_elem) &
+                ~ep_config->proto_lane_map;
     ucs_for_each_bit(lane, lane_map) {
         ucs_assertv(lane < UCP_MAX_LANES,
                     "lane=%" PRIu8 ", lane_map=0x%" PRIx16, lane, lane_map);
-        rsc_index = ep_config_key->lanes[lane].rsc_index;
-        wiface    = ucp_worker_iface(worker, rsc_index);
+        wiface = ucp_worker_iface_by_lane(worker, ep_config, lane);
```
Comment (brminich) [inline src/ucp/proto/proto_select.c:564]: can `ucp_ep_config_for_each_proto_lane` be used here as well?

**Review by yosefe — APPROVED** [review]

File: src/ucp/proto/proto_select.c (hunk 14 above)
Comment (rakhmets) [reply src/ucp/proto/proto_select.c]: Done.

**Review by brminich — APPROVED** [review]

**Pull request merged**